pandas/tests/series/test_analytics.py

```diff
@@ -237,6 +238,67 @@ def test_npdiff(self):
         r = np.diff(s)
         assert_series_equal(Series([nan, 0, 0, 0, nan]), r)
 
+    def test_int_diff(self):
+        # int dtype
+        a = 10000000000000000
+        b = a + 1
+        s = Series([a, b])
+
+        result = s.diff()
+        assert result[1] == 1
+
+    def test_tz_diff(self):
+        # Combined datetime diff, normal diff and boolean diff test
+        ts = tm.makeTimeSeries(name="ts")
+        ts.diff()
+
+        # neg n
+        result = ts.diff(-1)
+        expected = ts - ts.shift(-1)
+        assert_series_equal(result, expected)
+
+        # 0
+        result = ts.diff(0)
+        expected = ts - ts
+        assert_series_equal(result, expected)
+
+        # datetime diff (GH3100)
+        s = Series(date_range("20130102", periods=5))
+        result = s.diff()
+        expected = s - s.shift(1)
+        assert_series_equal(result, expected)
+
+        # timedelta diff
+        result = result - result.shift(1)  # previous result
+        expected = expected.diff()  # previously expected
+        assert_series_equal(result, expected)
+
+        # with tz
+        s = Series(
+            date_range("2000-01-01 09:00:00", periods=5, tz="US/Eastern"), name="foo"
+        )
+        result = s.diff()
+        expected = Series(TimedeltaIndex(["NaT"] + ["1 days"] * 4), name="foo")
+        assert_series_equal(result, expected)
+
+    @pytest.mark.parametrize(
+        "input,output,diff",
+        [([False, True, True, False, False], [nan, True, False, True, False], 1)],
+    )
+    def test_bool_diff(self, input, output, diff):
+        # boolean series (test for fixing #17294)
+        s = Series(input)
+        result = s.diff()
+        expected = Series(output)
+        assert_series_equal(result, expected)
```

Comment:
is x - x.shift() also true here (as it is for obj_diff)? can you add that assert as well (do another result= and expected=), otherwise this PR lgtm.

Reply:
No, it is not and will not with the current implementation. @TomAugspurger [specifically wanted a boolean Series](https://github.com/pandas-dev/pandas/pull/28251#discussion_r320818052) back instead of a list of 1,0,-1. As shown in the quote below.

> I'm having a hard time following, but IMO these two should have the same output
> 
> ```python
> In [3]: pd.Series([True, False, True]).diff()
> Out[3]:
> 0     NaN
> 1    True
> 2    True
> dtype: object
> 
> In [4]: pd.Series([True, False, True]).astype(object).diff()
> Out[4]:
> 0    NaN
> 1     -1
> 2      1
> dtype: object
> ```
> 
> I think both of those should be `Out[3]`, the same result as `np.diff` with the missing values handled appropriately.

Reply:
ok fair enough, do we have a followup issue on this? @TomAugspurger ? (which likely will block on a real BoolType EA)